### ISSUE TITLE
Support hot-reload / hot-restart using vm service protocol

### ISSUE THREAD

**Issue by fuzzybinary** [Dart-Code/Dart-Code#2708]
The Dart VM Service protocol now supports hot reload and hot restart, meaning Dart command line apps should support proper code hot reloading.  It would be nice if Dart-Code utilized this method instead of relying on the flutter tool.

**Comment by DanTup** [comment]
@bkonyi should there be something like a reloadSources service/extension registered when running a Dart script? I can see the VM Service has a `reloadServices` method, but in Flutter we use the presence of `reloadSources` and `hotRestart` services to know when these are available. Should there be similar for Dart, or should we assume it's always available for a recent SDK?

(Also, the original issue here mentions Restart too - but I don't see anything in the docs about that - is it also available through the Dart VM without Flutter?)

**Comment by fuzzybinary** [comment]
I may have been mistaken about the ability to do Hot Restart without Flutter.

**Comment by DanTup** [comment]
Ah, got it.

It would be good to support Hot Reload still - but it'd be nice to do it consistently with Flutter (eg. we use the presence of some VM Service to know it's available, and call it via that service) so the extension doesn't need to have logic about when it may/may not be available.

**Comment by bkonyi** [comment]
Correct, the service protocol doesn't actually implement Hot Restart. That's solely a Flutter extension of the protocol. The VM service protocol does support [`reloadSources`](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#reloadsources), which is the RPC to trigger a hot reload.

**Comment by DanTup** [comment]
Ah, sounds like there does need to be two paths:

- Dart: assume Hot Reload is always available, call `reloadSources`
- Flutter: use the presence of the `reloadSources` _Service_ to decide if it's available, and call that instead of `reloadSources` directly (I presume Flutter is doing something more than just calling the VM service)

The second is all already there, the first would just be adding a new **Dart: Hot Reload** command that is visible only when running a non-Flutter app, that just calls `reloadSources`.
